From 1c80f079b12e09355d646cf60f0143afed513d8c Mon Sep 17 00:00:00 2001
From: Radim Krcmar <rkrcmar@redhat.com>
Date: Wed, 12 Nov 2014 16:32:00 -0500
Subject: [virt] kvm/vmx: make PLE window per-VCPU

Message-id: <1415809922-10128-4-git-send-email-rkrcmar@redhat.com>
Patchwork-id: 99847
O-Subject: [RHEL7.1 kernel PATCH 3/5] KVM: VMX: make PLE window per-VCPU
Bugzilla: 1163296
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Andrew Jones <drjones@redhat.com>
RH-Acked-by: Paolo Bonzini <pbonzini@redhat.com>

Change PLE window into per-VCPU variable, seeded from module parameter,
to allow greater flexibility.

Brings in a small overhead on every vmentry.

Signed-off-by: Radim Krcmar <rkrcmar@redhat.com>
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
(cherry picked from commit a7653ecdf34c68a1af4fc085511afcf7ff011903)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kvm/vmx.c b/arch/x86/kvm/vmx.c
index 62a19bb..e45ba1c 100644
--- a/arch/x86/kvm/vmx.c
+++ b/arch/x86/kvm/vmx.c
@@ -481,6 +481,10 @@ struct vcpu_vmx {
 
  /* Support for a guest hypervisor (nested VMX) */
  struct nested_vmx nested;
+
+ /* Dynamic PLE window. */
+ int ple_window;
+ bool ple_window_dirty;
 };
 
 enum segment_cache_field {
@@ -4396,7 +4400,8 @@ static int vmx_vcpu_setup(struct vcpu_vmx *vmx)
 
  if (ple_gap) {
   vmcs_write32(PLE_GAP, ple_gap);
-  vmcs_write32(PLE_WINDOW, ple_window);
+  vmx->ple_window = ple_window;
+  vmx->ple_window_dirty = true;
  }
 
  vmcs_write32(PAGE_FAULT_ERROR_CODE_MASK, 0);
@@ -7388,6 +7393,11 @@ static void __noclone vmx_vcpu_run(struct kvm_vcpu *vcpu)
  if (vmx->emulation_required)
   return;
 
+ if (vmx->ple_window_dirty) {
+  vmx->ple_window_dirty = false;
+  vmcs_write32(PLE_WINDOW, vmx->ple_window);
+ }
+
  if (vmx->nested.sync_shadow_vmcs) {
   copy_vmcs12_to_shadow(vmx);
   vmx->nested.sync_shadow_vmcs = false;
-- 
1.7.1